### ISSUE TITLE
feat: SSD (Single Shot MultiBox Detector) による物体検出サンプルの追加

### ISSUE THREAD

**Issue by yukiharada1228** [yukiharada1228/KnowledgeTransferGraph#6]
## 概要

[MPRGDeepLearningLectureNotebook の SSD 実装](https://github.com/machine-perception-robotics-group/MPRGDeepLearningLectureNotebook/blob/master/11_cnn_pytorch/07_single_shot_multibox_detector.ipynb) を参考に、本フレームワーク (`dml`) を使った物体検出サンプルを `examples/VOC/SSD/` として追加する。

---

## 背景

現在のフレームワークは画像分類タスク（CIFAR-10/100）のみ対応している。  
SSD を実装することで **物体検出タスクへの対応** と、**複数 SSD モデル間での DML (Deep Mutual Learning)** が可能になる。

---

### 1. SSD300 モデル (`models/ssd300.py`)

- **バックボーン**: VGG-16（ImageNet 事前学習済み）
- **予測ヘッド**: Conv4_3, Conv7, Conv8_2, Conv9_2, Conv10_2, Conv11_2 の 6 スケール
- **L2 正規化**: Conv4_3 出力へ適用
- **デフォルトボックス生成**: 各特徴マップに対応した aspect ratio / scale の設定

### 2. SSD 損失関数 (`losses.py` または `examples/VOC/SSD/losses.py`)

参照ノートブックの `MYLoss` を基に実装：

- **分類損失**: クロスエントロピー + **Hard Negative Mining**（正例:負例 = 1:3）
- **位置推定損失**: Smooth L1 Loss（正例 box のみ適用）
- モデル出力は `(loc_preds, cls_preds)` のタプルを想定

```python
class SSDLoss(nn.Module):
    def forward(self, predictions, targets):
        loc_preds, cls_preds = predictions  # モデル出力
        # targets: List[Tensor] — 各画像のアノテーション [x1,y1,x2,y2,label]
        ...
```

### 3. データセット (`dataset.py`)

- **Pascal VOC 2007/2012** を使用（`torchvision.datasets.VOCDetection`）
- データ拡張: ランダムクロップ、水平反転、色変換、正規化
- `prepare_batch` を object detection 向けに拡張（可変長アノテーションのコレート）

### 4. 評価関数 (`eval.py`)

- **mAP (mean Average Precision)** の実装（IoU 閾値 0.5）
- `Node.eval_fn` として渡せるインタフェース: `(model, device) -> float`

### 5. 学習スクリプト

| ファイル | 説明 |
|---|---|
| `independent_train.py` | SSD 単体の教師なし学習 |
| `dml_train.py` | 2 つの SSD が互いに蒸留する DML 学習 |
| `kd_train.py` | 事前学習済み大規模 SSD を教師とした Knowledge Distillation |

### 6. フレームワーク側の拡張検討

| 課題 | 対応方針 |
|---|---|
| `Edge.compute` が分類ロジットを前提 | SSD 出力 `(loc, cls)` タプルをそのまま `criterion` へ渡す（`temperature=None` の既存パスで対応可能） |
| `prepare_batch` のデフォルトが `(image, label)` 前提 | `Trainer` に `prepare_batch` を渡すことで対応済み。検出用の実装をサンプル側で提供 |
| `Trainer._get_batch_size` がネストしたリストに未対応 | 必要に応じて `List[Tensor]` 対応を追加 |

---

## ディレクトリ構成（案）

```
examples/VOC/SSD/
├── README.md
├── dataset.py          # VOCDetection ラッパー + データ拡張
├── losses.py           # SSDLoss (分類 + 位置推定 + Hard Negative Mining)
├── eval.py             # mAP 評価
├── models/
│   └── ssd300.py       # SSD300 (VGG-16 ベース)
├── independent_train.py
├── dml_train.py
└── kd_train.py
```

---

## 参考資料

- 参照ノートブック: [07_single_shot_multibox_detector.ipynb](https://github.com/machine-perception-robotics-group/MPRGDeepLearningLectureNotebook/blob/master/11_cnn_pytorch/07_single_shot_multibox_detector.ipynb)
- 元論文: [SSD: Single Shot MultiBox Detector (Liu et al., 2016)](https://arxiv.org/abs/1512.02325)
- Pascal VOC データセット: `torchvision.datasets.VOCDetection`

---

## 完了条件

- [ ] `examples/VOC/SSD/` 以下に上記ファイルが揃っている
- [ ] `independent_train.py` が VOC2007 で mAP > 70% を達成
- [ ] `dml_train.py` で DML 学習が正常に動作する
- [ ] `README.md` に実行手順・精度結果が記載されている
